cmd/qat_plugin/dpdkdrv/dpdkdrv.go

```diff
@@ -245,56 +245,108 @@ func (dp *DevicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
 	return nil
 }
 
+func getPciDevicesWithPattern(pattern string) (pciDevices []string) {
+	pciDevices = make([]string, 0)
+
+	devs, err := filepath.Glob(pattern)
+	if err != nil {
+		klog.Warningf("bad pattern: %s", pattern)
+		return
+	}
+
+	for _, devBdf := range devs {
+		targetDev, err := filepath.EvalSymlinks(devBdf)
+		if err != nil {
+			klog.Warningf("unable to evaluate symlink: %s", devBdf)
+			continue
+		}
+		pciDevices = append(pciDevices, targetDev)
+	}
+
+	return
+}
+
+func (dp *DevicePlugin) getVfDevices() (qatVfDevices []string) {
+	qatPfDevices := make([]string, 0)
+	qatVfDevices = make([]string, 0)
+
+	// Get PF BDFs bound to a PF driver
+	for _, vfDriver := range dp.kernelVfDrivers {
+		pfDriver := strings.TrimSuffix(vfDriver, "vf")
+		pattern := filepath.Join(dp.pciDriverDir, pfDriver, "????:??:??.?")
+		qatPfDevices = append(qatPfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	// Get VF devices belonging to a PF device
+	for _, qatPfDevice := range qatPfDevices {
+		pattern := filepath.Join(qatPfDevice, "virtfn*")
+		qatVfDevices = append(qatVfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	if len(qatPfDevices) > 0 && len(qatVfDevices) > 0 {
+		return
+	}
+
+	// No PF devices found, running in a VM?
+	for _, vfDriver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
```

Comment:
Is the order of the drivers important here? I guess
```go
for _, vfDriver := range append(dp.kernelVfDrivers, dp.dpdkDriver) {
```
would be more readable.

Reply:
We prefer devices that are already bound to the targeted driver. See: #146